### PR TITLE
Cleanup allocation.

### PR DESCRIPTION
**Pull request by chriseth**

### CODE REVIEW

File: test/libsolidity/gasTests/abiv2_optimised.sol [hunk 1]
```diff
@@ -17,9 +17,9 @@ contract C {
 // optimize-yul: true
 // ----
 // creation:
-//   codeDepositCost: 587400
+//   codeDepositCost: 588800
```
Comment (chriseth) [inline test/libsolidity/gasTests/abiv2_optimised.sol]: It is weird - I checked the optimized yul code (from the new codegen) and it actually looks shorter to me. So the increase seems to be in the code emitted by the old codegen.

File: libsolidity/codegen/YulUtilFunctions.cpp [hunk 2]
```diff
@@ -2800,27 +2800,24 @@ string YulUtilFunctions::prepareStoreFunction(Type const& _type)
...
+	string functionName = "allocate_memory";
 	return m_functionCollector.createFunction(functionName, [&]() {
 		return Whiskers(R"(
 			function <functionName>(size) -> memPtr {
-				memPtr := mload(<freeMemoryPointer>)
-				let newFreePtr := add(memPtr, size)
-				// protect against overflow
-				if or(gt(newFreePtr, 0xffffffffffffffff), lt(newFreePtr, memPtr)) { <panic>() }
-				mstore(<freeMemoryPointer>, newFreePtr)
+				memPtr := <allocateUnbounded>()
```
Comment (axic) [inline libsolidity/codegen/YulUtilFunctions.cpp:2807]: The code above uses `mem`. Should we just use `memPtr` consistently?

File: libsolidity/codegen/YulUtilFunctions.cpp [hunk 3]
```diff
@@ -2800,27 +2800,24 @@ string YulUtilFunctions::prepareStoreFunction(Type const& _type)
...
 {
-	string functionName = "allocateMemory";
+	string functionName = "allocate_memory";
 	return m_functionCollector.createFunction(functionName, [&]() {
 		return Whiskers(R"(
 			function <functionName>(size) -> memPtr {
-				memPtr := mload(<freeMemoryPointer>)
-				let newFreePtr := add(memPtr, size)
-				// protect against overflow
-				if or(gt(newFreePtr, 0xffffffffffffffff), lt(newFreePtr, memPtr)) { <panic>() }
```
Comment (axic) [inline libsolidity/codegen/YulUtilFunctions.cpp:2810]: Where is this handled now? Within finalize?

File: libsolidity/codegen/YulUtilFunctions.cpp [hunk 4]
```diff
@@ -2311,16 +2311,16 @@ string YulUtilFunctions::copyArrayFromStorageToMemoryFunction(ArrayType const& _
 				function <functionName>(slot) -> memptr {
 					memptr := <allocateTemp>()
 					let end := <encode>(slot, memptr)
-					mstore(<freeMemoryPointer>, end)
+					<finalize>(memptr, sub(end, memptr))
 				}
 			)")
 			("functionName", functionName)
-			("allocateTemp", allocationTemporaryMemoryFunction())
+			("allocateTemp", allocateUnboundedFunction())
```
Comment (axic) [inline libsolidity/codegen/YulUtilFunctions.cpp]: In other places it is using `allocateUnbounded`. Would be nice to use consistent naming.

File: libsolidity/codegen/YulUtilFunctions.cpp [hunk 5]
```diff
@@ -2311,16 +2311,16 @@ string YulUtilFunctions::copyArrayFromStorageToMemoryFunction(ArrayType const& _
...
 			(
 				"encode",
 				abi.abiEncodeAndReturnUpdatedPosFunction(_from, _to, ABIFunctions::EncodingOptions{})
 			)
-			("freeMemoryPointer", to_string(CompilerUtils::freeMemoryPointer))
+			("finalize", finalizeAllocationFunction())
```
Comment (axic) [inline libsolidity/codegen/YulUtilFunctions.cpp]: I think `finalizeAllocation` would be better throughout. With `finalize` one needs to look up what it means.

File: libsolidity/codegen/YulUtilFunctions.cpp [hunk 6]
```diff
@@ -2833,12 +2830,32 @@ string YulUtilFunctions::allocationTemporaryMemoryFunction()
...
+		)")
+		("functionName", functionName)
+		("freeMemoryPointer", to_string(CompilerUtils::freeMemoryPointer))
+		("roundUp", roundUpFunction())
+		("panic", panicFunction(PanicCode::ResourceError))
+		.render();
+	});
+}
+
+string YulUtilFunctions::releaseMemoryFunction()
```
Comment (axic) [inline libsolidity/codegen/YulUtilFunctions.cpp]: Do we still need this?

File: libsolidity/codegen/YulUtilFunctions.cpp [hunk 7]
```diff
@@ -4219,11 +4236,7 @@ string YulUtilFunctions::extractReturndataFunction()
 					}
 					default {
 						// allocate some memory into data of size returndatasize() + PADDING
-						data := <allocate>(<roundUp>(add(returndatasize(), 0x20)))
-
-						// store array length into the front
-						mstore(data, returndatasize())
-
+						data := <allocateArray>(returndatasize())
```
Comment (axic) [inline libsolidity/codegen/YulUtilFunctions.cpp]: Perhaps we could merge this change separately? It seems to be not related?

File: libsolidity/codegen/YulUtilFunctions.cpp [hunk 8]
```diff
@@ -2833,12 +2830,32 @@ string YulUtilFunctions::allocationTemporaryMemoryFunction()
...
+
+string YulUtilFunctions::releaseMemoryFunction()
 {
-	string functionName = "releaseTemporaryMemory";
+	string functionName = "release_memory";
 	return m_functionCollector.createFunction(functionName, [&](){
 		return Whiskers(R"(
-			function <functionName>() {
+			function <functionName>(ptr) {
 			}
```
Comment (axic) [inline libsolidity/codegen/YulUtilFunctions.cpp]: Shouldn't technically this have an assertion for `ptr == freeMemoryPointer`?

File: libsolidity/codegen/ir/IRGeneratorForStatements.cpp [hunk 9]
```diff
@@ -1105,8 +1105,11 @@ void IRGeneratorForStatements::endVisit(FunctionCall const& _functionCall)
 			else
 			{
 				// Used to reset the free memory pointer later.
+				// This is an abuse of the `allocateUnbounded` function.
```
Comment (axic) [inline libsolidity/codegen/ir/IRGeneratorForStatements.cpp]: Add a TODO here so it is easier to find.

File: libsolidity/codegen/ir/IRGeneratorForStatements.cpp [hunk 10]
```diff
@@ -1122,25 +1125,25 @@ void IRGeneratorForStatements::endVisit(FunctionCall const& _functionCall)
 					"))\n";
 				IRVariable selectorVariable(m_context.newYulVariable(), *TypeProvider::fixedBytes(4));
 				define(selectorVariable, hashVariable);
-				m_code << "mstore(" << to_string(CompilerUtils::freeMemoryPointer) << ", " << freeMemoryPre << ")\n";
+				m_code << m_utils.allocateUnboundedFunction() << "(" << freeMemoryPre << ", 0)\n";
```
Comment (axic) [inline libsolidity/codegen/ir/IRGeneratorForStatements.cpp]: Hmm? `allocateUnbounded` does not seem to take two arguments? And the `mstore` is gone.

File: libsolidity/codegen/ir/IRGeneratorForStatements.cpp [hunk 11]
```diff
@@ -1403,17 +1405,17 @@ void IRGeneratorForStatements::endVisit(FunctionCall const& _functionCall)
 			<!saltSet>
 				let <address> := create(<value>, <memPos>, sub(<memEnd>, <memPos>))
 			</saltSet>
+			<releaseMemory>(<memPos>)
```
Comment (axic) [inline libsolidity/codegen/ir/IRGeneratorForStatements.cpp]: This looks like a bugfix, doesn't it?

Shouldn't we handle this separately and/or add tests for this?

File: libsolidity/codegen/ir/IRGeneratorForStatements.cpp [hunk 12]
```diff
@@ -1486,19 +1488,21 @@ void IRGeneratorForStatements::endVisit(FunctionCall const& _functionCall)
...
 		}
 		Whiskers templ(R"(
-			let <pos> := <allocateTemporary>()
+			let <pos> := <allocateUnbounded>()
 			let <end> := <encodeArgs>(<pos> <argumentString>)
 			<?isECRecover>
 				mstore(0, 0)
 			</isECRecover>
 			let <success> := <call>(<gas>, <address> <?isCall>, 0</isCall>, <pos>, sub(<end>, <pos>), 0, 32)
+			<releaseMemory>(<pos>)
```
Comment (axic) [inline libsolidity/codegen/ir/IRGeneratorForStatements.cpp]: Same question here.

File: libsolidity/codegen/YulUtilFunctions.cpp (hunk 3 above)
Comment (chriseth) [reply libsolidity/codegen/YulUtilFunctions.cpp:2810]: yes.

File: libsolidity/codegen/YulUtilFunctions.cpp (hunk 6 above)
Comment (chriseth) [reply libsolidity/codegen/YulUtilFunctions.cpp]: Yes.

File: libsolidity/codegen/YulUtilFunctions.cpp (hunk 8 above)
Comment (chriseth) [reply libsolidity/codegen/YulUtilFunctions.cpp]: This is unused functionality. We might decide to write a complicated allocator ...

File: libsolidity/codegen/ir/IRGeneratorForStatements.cpp (hunk 10 above)
Comment (chriseth) [reply libsolidity/codegen/ir/IRGeneratorForStatements.cpp]: A test coverage gap!

File: libsolidity/codegen/ir/IRGeneratorForStatements.cpp (hunk 11 above)
Comment (chriseth) [reply libsolidity/codegen/ir/IRGeneratorForStatements.cpp]: `releaseMemory` is a no-op.

File: libsolidity/codegen/YulUtilFunctions.cpp (hunk 2 above)
Comment (chriseth) [reply libsolidity/codegen/YulUtilFunctions.cpp:2807]: Whereis 'mem' used?

Comment (axic) [reply libsolidity/codegen/YulUtilFunctions.cpp:2807]: In the case above this.

Comment (axic) [reply libsolidity/codegen/YulUtilFunctions.cpp:2807]: Also found `mpos` being used in places, like `YulUtilFunctions::externalCodeFunction`.

**Comment by chriseth** [comment]
Rebased.

File: libsolidity/codegen/YulUtilFunctions.cpp [hunk 13]
```diff
@@ -4175,13 +4191,13 @@ string YulUtilFunctions::tryDecodeErrorMessageFunction()
 				let end := add(add(msg, 0x20), length)
 				if gt(end, add(data, sub(returndatasize(), 4))) { leave }
 
-				mstore(<freeMemoryPointer>, add(add(msg, 0x20), <roundUp>(length)))
+				<finalizeAllocation>(data, add(add(offset, msg), add(0x20, length)))
```
Comment (cameel) [inline libsolidity/codegen/YulUtilFunctions.cpp]: This looks wrong. `add(offset, msg)` equals `offset + data + offset`. I think that `offset` should be there only once and also `data` should not be passed into the `size` parameter of `finalizeAllocation()`.

Old version was `fmp = data + offset + 0x20 + roundUp(length)`.
The new version is basically `fmp = data + offset + data + offset + 0x20 + length`.

```suggestion
				<finalizeAllocation>(data, add(offset, add(0x20, length)))
```

Comment (cameel) [reply libsolidity/codegen/YulUtilFunctions.cpp]: Also, was `roundUp()` removed on purpose?

File: libsolidity/codegen/ir/IRGeneratorForStatements.cpp (hunk 10 above)
Comment (axic) [reply libsolidity/codegen/ir/IRGeneratorForStatements.cpp]: Did you add a test case for this?

File: libsolidity/codegen/YulUtilFunctions.cpp (hunk 13 above)
Comment (chriseth) [reply libsolidity/codegen/YulUtilFunctions.cpp]: roundup is now inside finalizeAllocation

Comment (cameel) [reply libsolidity/codegen/YulUtilFunctions.cpp]: If this passed tests then this is apparently not covered. Is something like this testable via inline assembly?

File: libsolidity/codegen/YulUtilFunctions.h [hunk 14]
```diff
@@ -352,16 +352,25 @@ class YulUtilFunctions
...
-	/// Return value: pointer
+	/// signature: (size) -> ptr
 	std::string allocationFunction();
 
-	/// @returns the name of the function that allocates temporary memory with predefined size
+	/// @returns the name of the function that allocates memory whose size will be defined later,
+	/// or it will be released. The allocation can be finalized using finalizeAllocationFunction.
+	/// Any other allocation will invalidate the memory pointer unless finalizeAllocationFunction
+	/// is called.
 	/// Return value: pointer
```
Comment (cameel) [inline libsolidity/codegen/YulUtilFunctions.h]: Return value was removed in other comments so I think it should be removed here too.
```suggestion
```

File: libsolidity/codegen/YulUtilFunctions.h [hunk 15]
```diff
@@ -352,16 +352,25 @@ class YulUtilFunctions
...
+	/// @returns the name of the function that finalizes an unbounded memory allocation,
+	/// i.e. sets its size.
+	/// signature: (ptr, size) ->
+	std::string finalizeAllocationFunction();
 
-	/// @returns the name of the function that releases previously allocated temporary memory
-	std::string releaseTemporaryMemoryFunction();
+	/// @returns the name of the function that releases previously allocated memory
+	/// signature: (ptr) ->
+	std::string releaseMemoryFunction();
```
Comment (cameel) [inline libsolidity/codegen/YulUtilFunctions.h]: Can you release memory allocated by both `allocate()` and `allocate_unbounded()` or only by the former? `allocate_unbounded()` does not even modify the free memory pointer so I'd guess you can't release memory allocated by it but I see in the diff that now `endVisit(FunctionCall)` does just that.

This should be clarified in the docstring because this is pretty arbitrary and depends on how you expect `release_memory()` to work in the future.

File: libsolidity/codegen/YulUtilFunctions.cpp [hunk 16]
```diff
@@ -2309,18 +2309,18 @@ string YulUtilFunctions::copyArrayFromStorageToMemoryFunction(ArrayType const& _
 			ABIFunctions abi(m_evmVersion, m_revertStrings, m_functionCollector);
 			return Whiskers(R"(
 				function <functionName>(slot) -> memptr {
-					memptr := <allocateTemp>()
+					memptr := <allocateUnbounded>()
```
Comment (cameel) [inline libsolidity/codegen/YulUtilFunctions.cpp]: For consistency:
```suggestion
					memPtr := <allocateUnbounded>()
```

**Review by cameel — DISMISSED** [review]
I think that https://github.com/ethereum/solidity/pull/10769#discussion_r559756949 is pretty important and the rules for when using release is allowed should be documented in the docstring. 

These are seem easy to fix though and I don't see any other problems so this PR should be ready to merge as soon as these are done.

File: libsolidity/codegen/YulUtilFunctions.cpp (hunk 8 above)
Comment (axic) [reply libsolidity/codegen/YulUtilFunctions.cpp]: I wish there would be a way to have this ensured in a no-overhead version. What is the point having the hypothetical free memory when it is not tested.

File: libsolidity/codegen/ir/IRGeneratorForStatements.cpp [hunk 17]
```diff
@@ -1129,18 +1132,18 @@ void IRGeneratorForStatements::endVisit(FunctionCall const& _functionCall)
 			selector = convert(*arguments.front(), *TypeProvider::fixedBytes(4)).name();
 
 		Whiskers templ(R"(
-			let <data> := <allocateTemporary>()
+			let <data> := <allocateUnbounded>()
 			let <mpos> := add(<data>, 0x20)
```
Comment (axic) [inline libsolidity/codegen/ir/IRGeneratorForStatements.cpp]: `mpos` to `memPtr` for consistency?

File: libsolidity/codegen/YulUtilFunctions.cpp (hunk 2 above)
Comment (axic) [reply libsolidity/codegen/YulUtilFunctions.cpp:2807]: > Also found mpos being used in places, like YulUtilFunctions::externalCodeFunction.

I think this was not addressed yet.

Comment (cameel) [reply libsolidity/codegen/YulUtilFunctions.cpp:2807]: Sorry for that, most of other comments were addressed so I went over this one too quickly and only checked for `mem` :)

File: libsolidity/codegen/YulUtilFunctions.h (hunk 15 above)
Comment (chriseth) [reply libsolidity/codegen/YulUtilFunctions.h]: Maybe it's better to remove the memory release function again.

File: libsolidity/codegen/YulUtilFunctions.cpp (hunk 13 above)
Comment (chriseth) [reply libsolidity/codegen/YulUtilFunctions.cpp]: It reserved more memory than needed. Maybe when adding gas cost tests, we could also try to check the final value of the free memory pointer somehow?

**Comment by chriseth** [comment]
Removed the release function, updated the documentation in the header and tried to consistently name some of the local variables.

File: test/libsolidity/gasTests/abiv2.sol [hunk 18]
```diff
@@ -14,9 +14,9 @@ contract C {
 }
 // ----
 // creation:
-//   codeDepositCost: 1175600
-//   executionCost: 1221
-//   totalCost: 1176821
+//   codeDepositCost: 1181400
+//   executionCost: 1227
+//   totalCost: 1182627
```
Comment (bshastry) [inline test/libsolidity/gasTests/abiv2.sol:19]: NTS: Gas increase by about 6k

File: libsolidity/codegen/YulUtilFunctions.cpp [hunk 19]
```diff
@@ -2800,28 +2800,24 @@ string YulUtilFunctions::prepareStoreFunction(Type const& _type)
 
 string YulUtilFunctions::allocationFunction()
 {
-	string functionName = "allocateMemory";
+	string functionName = "allocate_memory";
```
Comment (bshastry) [inline libsolidity/codegen/YulUtilFunctions.cpp:2803]: Why underscore instead of camel case?

File: libsolidity/codegen/YulUtilFunctions.cpp [hunk 20]
```diff
@@ -183,15 +183,15 @@ string YulUtilFunctions::requireOrAssertFunction(bool _assert, Type const* _mess
 		return Whiskers(R"(
 			function <functionName>(condition <messageVars>) {
 				if iszero(condition) {
-					let fmp := mload(<freeMemPointer>)
-					mstore(fmp, <errorHash>)
-					let end := <abiEncodeFunc>(add(fmp, <hashHeaderSize>) <messageVars>)
-					revert(fmp, sub(end, fmp))
+					let memPtr := <allocateUnbounded>()
```
Comment (bshastry) [inline libsolidity/codegen/YulUtilFunctions.cpp:186]: Actually `allocateUnbounded` function is not performing the allocation but simply placing the `mload(fmp)` into a function called `allocateUnbounded`. To me, a function name such as `obtainFreeMemoryPtr` seems more appropriate.

Comment (bshastry) [reply libsolidity/codegen/YulUtilFunctions.cpp:186]: Also, the renaming of `fmp` to `memPtr` feels unnecessary since the semantics are unchanged.

File: libsolidity/codegen/YulUtilFunctions.cpp [hunk 21]
```diff
@@ -2834,15 +2830,22 @@ string YulUtilFunctions::allocationTemporaryMemoryFunction()
...
+string YulUtilFunctions::finalizeAllocationFunction()
 {
-	string functionName = "releaseTemporaryMemory";
-	return m_functionCollector.createFunction(functionName, [&](){
+	string functionName = "finalize_allocation";
+	return m_functionCollector.createFunction(functionName, [&]() {
 		return Whiskers(R"(
-			function <functionName>() {
+			function <functionName>(memPtr, size) {
+				let newFreePtr := add(memPtr, <roundUp>(size))
```
Comment (bshastry) [inline libsolidity/codegen/YulUtilFunctions.cpp:2839]: Does `size` need to be rounded up even if it is already a multiple of 32?

**Review by bshastry — COMMENTED** [review]
Minor comments

File: libsolidity/codegen/YulUtilFunctions.cpp (hunk 20 above)
Comment (cameel) [reply libsolidity/codegen/YulUtilFunctions.cpp:186]: I think the idea here is that you should no longer be able to freely touch free memory pointer and should treat it as an implementation detail of the the allocate/release functions instead. `allocateUnbounded()` does not do anything with it currently but this might change.

File: libsolidity/codegen/YulUtilFunctions.cpp (hunk 19 above)
Comment (chriseth) [reply libsolidity/codegen/YulUtilFunctions.cpp:2803]: Yul util functions all use snake case.

File: libsolidity/codegen/YulUtilFunctions.cpp (hunk 21 above)
Comment (chriseth) [reply libsolidity/codegen/YulUtilFunctions.cpp:2839]: If it is already a multiple of 32, it will not change.

File: libsolidity/codegen/YulUtilFunctions.cpp (hunk 20 above)
Comment (chriseth) [reply libsolidity/codegen/YulUtilFunctions.cpp:186]: Sure, this is mostly perceived meaning. The main idea of this PR is to abstract away the free memory pointer in general. It now only occurs in the functions related to allocation.

File: libsolidity/codegen/YulUtilFunctions.h (hunk 15 above)
Comment (cameel) [reply libsolidity/codegen/YulUtilFunctions.h]: Is the plan to reintroduce it some point though?

Comment (cameel) [reply libsolidity/codegen/YulUtilFunctions.h]: I also wonder if we might want to preserve the distinction between the normal and temporary allocations. The latter were already paired (correctly - I'm assuming) with deallocations.

Comment (chriseth) [reply libsolidity/codegen/YulUtilFunctions.h]: Maybe, but I think it is rather complicated anyway.

**Review by cameel — COMMENTED** [review]
It looks fine to me now and I have nothing to add but there are some points raised by others so I'm not marking as approved yet.

File: libsolidity/codegen/YulUtilFunctions.cpp (hunk 21 above)
Comment (bshastry) [reply libsolidity/codegen/YulUtilFunctions.cpp:2839]: Ah I see, thanks!

File: libsolidity/codegen/YulUtilFunctions.cpp (hunk 2 above)
Comment (chriseth) [reply libsolidity/codegen/YulUtilFunctions.cpp:2807]: Can we keep unrelated code out of context of the "cleanup allocation functions" PR?

Comment (axic) [reply libsolidity/codegen/YulUtilFunctions.cpp:2807]: Sure, you started to rename them, didn't you? :)

If there's renaming, do it properly in one sweep. I don't care too much about the local variables names, only about the allocate/finalize names to be not misleading.

Comment (chriseth) [reply libsolidity/codegen/YulUtilFunctions.cpp:2807]: The stack element of a variable of memory reference type is called `mpos`, so I don't think it makes sense to rename all local variables called `mpos`.

File: libsolidity/codegen/ir/IRGeneratorForStatements.cpp (hunk 10 above)
Comment (chriseth) [reply libsolidity/codegen/ir/IRGeneratorForStatements.cpp]: Test is here: https://github.com/ethereum/solidity/pull/10776

File: libsolidity/codegen/YulUtilFunctions.cpp (hunk 20 above)
Comment (bshastry) [reply libsolidity/codegen/YulUtilFunctions.cpp:186]: Note to self: This change is superficial.

File: libsolidity/codegen/YulUtilFunctions.cpp [hunk 22]
```diff
@@ -2308,19 +2308,19 @@ string YulUtilFunctions::copyArrayFromStorageToMemoryFunction(ArrayType const& _
...
-					mstore(<freeMemoryPointer>, end)
+				function <functionName>(slot) -> memPtr {
+					memPtr := <allocateUnbounded>()
+					let end := <encode>(slot, memPtr)
+					<finalizeAllocation>(memPtr, sub(end, memPtr))
 				}
 			)")
 			("functionName", functionName)
-			("allocateTemp", allocationTemporaryMemoryFunction())
+			("allocateUnbounded", allocateUnboundedFunction())
```
Comment (bshastry) [inline libsolidity/codegen/YulUtilFunctions.cpp:2318]: Note to self: `allocateTemporaryMemoryFunction` is renamed to `allocateUnboundedFunction`

File: libsolidity/codegen/YulUtilFunctions.cpp [hunk 23]
```diff
@@ -2308,19 +2308,19 @@ string YulUtilFunctions::copyArrayFromStorageToMemoryFunction(ArrayType const& _
...
 			(
 				"encode",
 				abi.abiEncodeAndReturnUpdatedPosFunction(_from, _to, ABIFunctions::EncodingOptions{})
 			)
-			("freeMemoryPointer", to_string(CompilerUtils::freeMemoryPointer))
+			("finalizeAllocation", finalizeAllocationFunction())
```
Comment (bshastry) [inline libsolidity/codegen/YulUtilFunctions.cpp:2323]: Note to self: `finalizeAllocationFunction` essentially returns a function lambda that adds its arguments and stores the aligned addition in freeMemPtr.

File: libsolidity/codegen/YulUtilFunctions.cpp [hunk 24]
```diff
@@ -2308,19 +2308,19 @@ string YulUtilFunctions::copyArrayFromStorageToMemoryFunction(ArrayType const& _
...
 			ABIFunctions abi(m_evmVersion, m_revertStrings, m_functionCollector);
 			return Whiskers(R"(
-				function <functionName>(slot) -> memptr {
-					memptr := <allocateTemp>()
-					let end := <encode>(slot, memptr)
-					mstore(<freeMemoryPointer>, end)
+				function <functionName>(slot) -> memPtr {
+					memPtr := <allocateUnbounded>()
+					let end := <encode>(slot, memPtr)
+					<finalizeAllocation>(memPtr, sub(end, memPtr))
```
Comment (bshastry) [inline libsolidity/codegen/YulUtilFunctions.cpp:2314]: Why did `sub(end, memPtr)` replace `end`? Are we storing the length instead of absolute offset in freeMemPtr now?

File: libsolidity/codegen/YulUtilFunctions.cpp [hunk 25]
```diff
@@ -4175,13 +4178,13 @@ string YulUtilFunctions::tryDecodeErrorMessageFunction()
 				let end := add(add(msg, 0x20), length)
 				if gt(end, add(data, sub(returndatasize(), 4))) { leave }
 
-				mstore(<freeMemoryPointer>, add(add(msg, 0x20), <roundUp>(length)))
+				<finalizeAllocation>(data, add(offset, add(0x20, length)))
```
Comment (bshastry) [inline libsolidity/codegen/YulUtilFunctions.cpp:4181]: Not sure why we are replacing `mstore(freeMemPtr, msg + 0x20 + round(length))` by `mstore(data + round(offset + 0x20 + length)`

**Review by bshastry — COMMENTED** [review]
Some more comments

File: libsolidity/codegen/YulUtilFunctions.cpp (hunk 25 above)
Comment (cameel) [reply libsolidity/codegen/YulUtilFunctions.cpp:4181]: Because `finalize_allocation()` now manages `freeMemPtr`. The code is equivalent because `msg` is equal to `data + offset`, and `finalize_allocation()` already does `roundUp()`. The only difference is that `finalize_allocation()` also checks that the allocated memory area isn't too large (which probably accounts for some of the increased gas usage).

File: libsolidity/codegen/YulUtilFunctions.cpp (hunk 24 above)
Comment (cameel) [reply libsolidity/codegen/YulUtilFunctions.cpp:2314]: `mstore(<freeMemoryPointer>, end)` stores the new value of the free memory pointer, i.e. a memory location.`finalizeAllocation` on the other hand takes the current value of the free memory pointer and the size of the allocated memory. Internally it adds the two and passes them to `mstore()`.

**Review by mijovic — APPROVED** [review]
It looks good to me. Looks like all review comments were fixed.
There is probably more renaming to be done, but it is not a topic of this PR :)

**Review by cameel — APPROVED** [review]
